### ISSUE TITLE
Highcharts.correctFloat

### ISSUE THREAD

**Issue by juliusfriedman** [highcharts/highcharts#13873]
#### Expected behaviour
Chart library does not freeze or cause OOM

#### Actual behaviour
Browser freeze and OOM occurs.

#### Live demo with steps to reproduce
Render a `Histogram Chart` etc which gets stuck in the following stack:

![image](https://user-images.githubusercontent.com/7851618/87099809-b9c63600-c218-11ea-8d4d-f7aeb1f3f27e.png)

Seems to be stuck in `correctFloat`.

I have attempted to the following:

```ts
(function (H) {
  //@ts-ignore
  H.wrap(H.seriesTypes.histogram.prototype, 'setDerivedData', function (proceed) {
    debugger;
    if (this.baseSeries.yData.length !== 0 && this.baseSeries.xData.length !== 0) {
      proceed.apply(this)
    }
  })
  H.correctFloat = function (a, b) {
    debugger;
    return parseFloat(a.toPrecision(b));
  };
})(Highcharts)
```

No matter what I seem to do the OOM occurs in correctFloat and I cannot seem to override or replace it even with my code above OR additional code which does:
```ts
//@ts-ignore
Highcharts.correctFloat = function (a, b) {
    return
  };
```
#### Product version
8.1.2

#### Affected browser(s)
Chrome Version 83.0.4103.116 (Official Build) (64-bit)

**Comment by pawelfus** [comment]
Hi @juliusfriedman 

Thank you for reporting the issue. Could you provide live demo when it happens? Our demo seems to be unaffected: https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/demo/histogram

**Comment by juliusfriedman** [comment]
in the interim while I try to get something I share that can reproduce, Can you explain how one would override `correctFloat`?

**Comment by pawelfus** [comment]
> in the interim while I try to get something I share that can reproduce, Can you explain how one would override correctFloat?

It's not designed to overwrite, sorry.

> Also your example is a little weird, all the numbers have only 1 point after the decimal, is this an indication I should be rounding my output to 1 significant digit?

This is just a demo. Of course, you should avoid using 30 decimal digits in JS.. :)

**Comment by juliusfriedman** [comment]
> > in the interim while I try to get something I share that can reproduce, Can you explain how one would override correctFloat?
> 
> It's not designed to overwrite, sorry.
> 
> > Also your example is a little weird, all the numbers have only 1 point after the decimal, is this an indication I should be rounding my output to 1 significant digit?
> 
> This is just a demo. Of course, you should avoid using 30 decimal digits in JS.. :)

I agree on the precision anything over 15 digits is likely not that useful except in certain applications.

I do not agree on the `correctFloat` behavior, you either NEED to have a way to specify the max precision (param b) from `Highcharts.setOptions` or you must allow for a custom `correctFloat` to be injected.

This is needed regardless of if you can replicate this issue or not.

**Comment by juliusfriedman** [comment]
Would also be nice to have knobs for faster rendering e.g. when you notice your taking longer than X milliseconds to render what you have along with a loading indicator and do smaller chunks in intervals without blocking until the chart is loaded. 

I will admit some of the problem here is `angular-highcharts` and not `highcharts` specifically but none the less the API starts here especially for the other libraries, thus `maxRenderTimeMsec` is probably also a good option which would specify how long you have in each interval to render what you can.

**Comment by juliusfriedman** [comment]
https://jsfiddle.net/6yjg3527/

this is a problem with only 1000 elements let alone 10,000 or more.

Is there a max number of elements in highcharts one should consider? If not I would propose to portion your drawing so the blocking is not as bad.

**Comment by juliusfriedman** [comment]
@pawelfus let me know if anything needs clarification atp and thank you for your time.

**Comment by pawelfus** [comment]
> This is needed regardless of if you can replicate this issue or not.

I'm not sure why you need this. This is our internal method to correct numeric errors in JS, e.g. `0.1 + 0.2 != 0.3`. Could you elaborate? Thanks!

> Is there a max number of elements in highcharts one should consider?

There are no limits (only your machine limits - RAM, CPU etc.). If you want to load huge amount of points (e.g. 1 million points), then consider using boost module. It's designed to handle it. Rendering is done in canvas, instead of SVG. Personally, max number elements depends on a chart type. For example scatter 100k is simply unreadable: https://jsfiddle.net/BlackLabel/w4ua6j8n/1/ - human perception has it's limits too. More about boost module: https://www.highcharts.com/docs/advanced-chart-features/boost-module

**Comment by juliusfriedman** [comment]
Thank you for the information on the boost module.. I will give that a try.

I need this because it is only just, if I know that I am giving back values of a precision that I do not want to observe I should be able to at the client to tell it (since it is always going to call this function) what the max level of precision should be.

I see you didn't respond about 2), I suppose this is where the boost module will help..

I will give that a shot.

Thank you for your time and let me know if I can provide any further information.

**Comment by juliusfriedman** [comment]
Testing the boost was not straightforward, it is only enabled on certain chart types and unfortunately for some of the chart types we are using (arearange) it is not even supported: [See Caveats](https://www.highcharts.com/docs/advanced-chart-features/boost-module#caveats)

With that being typed, it does improve rendering speed with orders of magnitude for the charts where it is supported.

Finally the only other feedback I can provide is that some of the caveats listed seem lazy e.g. no triangles? no stroked line?

Why are not all charts supported? 

Also I never see a Canvas rendered, I would expect that to trump the performance of SVG as you indicated for a high number of elements.

Please also see your examples (which are not loading correctly in Chome due to missing `console.time` api's):

[Area Range Demo (But not supported?)](https://jsfiddle.net/gh/get/jquery/1.7.2/highslide-software/highcharts.com/tree/master/samples/highcharts/boost/arearange)

They do not seem to use canvas at all.

Example boost options I have tried:

```{
    enabled: true,
    useGPUTranslations: true,
    usePreallocated: true,
    seriesThreshold: 1,
    allowForce: true
  };```

**Comment by pawelfus** [comment]
Yes, not all series are supported in Boost module. Where did you find information that arearange is not supported? It's bug in docs, we should fix that.

> Why are not all charts supported?

Because not every series requires this support (e.g. pie chart with one million points). Some other series types are simply not implemented yet.

> Please also see your examples (which are not loading correctly in Chome due to missing console.time api's):

Demos works fine for me. I'm using:
> Google Chrome is up to date
> Version 83.0.4103.116 (Official Build) (64-bit)

And console.time API seems to be supported: https://developer.mozilla.org/en-US/docs/Web/API/Console/time

> In closing, the chart I had originally opened this issue for only had 1 data point on an arearange and was stuck in the correctFloat function for an unknown reason.

And this is the demo I'm requesting from the beginning. Please provide one. Example config is nice to have, but it's not enough, for example this data works fine: https://jsfiddle.net/BlackLabel/fgchamxj/

> I need this because it is only just, if I know that I am giving back values of a precision that I do not want to observe I should be able to at the client to tell it (since it is always going to call this function) what the max level of precision should be.

I don't remember why precision is set to 14 (max in JS is ~16, actually it's 15.95 digits to use safely - and 14 was set as default ~9years ago) - probably due to old browsers bugs, or we just assumed decimals in our charts most likely will start with leading digit other than 0. Number in JS is represented as 64bits (double precision), but fractions have only 52bits, thus value e.g. `0.12345678901234567890` will be displayed as `0.12345678901234568` (missing `90`). Adding integers e.g. `12.12345678901234567890` will cut off even more: `12.123456789012346` etc. etc. I'm not sure what you expect from us, but it seems you need to explain to your users how JS numbers work and how precision is handled..

Anyway, I still don't see a reason to make `correctFloat()` available in API (with an option to overwrite). We have various options to configure what you want to display to your end users, e.g. `formatter` and `format` options:
- https://api.highcharts.com/highcharts/tooltip.formatter
- https://api.highcharts.com/highcharts/series.line.dataLabels.format (+ formatting: https://www.highcharts.com/docs/chart-concepts/labels-and-string-formatting)
- ...

**Comment by juliusfriedman** [comment]
That is not a bug in the docs, the grid lines are not drawn for `arearange` charts I have confirmed this and `Triangles` are not drawn at all.

I will look into the `formatter` and `binwidth` as that is what `setDerived` data is doing.

I would highly suggest allowing the maxPrecision to be set reguardless.

Thank you for your time and I will see about getting you an example of the arearange / etc charts which have problems.

I will attach their config etc to this thread and if needed we can make new issues from there.

Thank you for your time!
